design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,201 @@
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+## Test plan
+
+This functionality is expected to work both in Interactive as well as Headless modes, and therefore
+all tests described below should be done in both:
+
+* Test that values can be serialized into JSON/AVRO/DELIMITED
+* Test that schemas are properly verified
+* Test `NULL` and table tombstones
+* Test `CREATE STREAM/TABLE` with topics that don't exist
+* Test producing data in topics that already exist
+* Test producing data to topics that are not associated with a KSQL source
+
+## Documentation Updates
+
+* The `CREATE TABLE` and `CREATE STREAM` syntax references will have an updated table properties
+section that includes the following rows:
+> ```rst
+> +=========================+======================================================================================================+
+> | KAFKA_TOPIC (required)  | The name of the Kafka topic that backs this source. The topic must either already exist in Kafka, or |
+> |                         | PARTITIONS and REPLICAS must be specified to create the topic. Command will fail if the topic exists |
+> |                         | with different partition/replica counts.                                                             |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> | PARTITIONS              | The number of partitions in the backing topic. This property must be set if creating a SOURCE without|
+> |                         | an existing topic.                                                                                   |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> | REPLICAS                | The number of replicas in the backing topic. This property must be set if creating a SOURCE without  |
+> |                         | an existing topic.                                                                                   |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> ```
+* A new section for `INSERT INTO ... VALUES` will be added to the `sytnax.rst`:
+> ```rst
+> **Synopsis**
+> 
+> .. code:: sql
+>
+>   INSERT INTO stream_name/table_name [( column_name [, ...] )]
+>     VALUES (value [, ...])
+> 
+> **Description**
+> 
+> Produce a row into an existing stream or table and its underlying topic based on
+> explicitly sepcified values.
+>
+> Column values can be given in several ways:
+> * Any column not explicitly given a value is set to its default (explicit or implicit) value.
+> * If no columns are specified, a value for every column is expected in the same order as the
+> schema
+> * To produce a ``NULL`` value, specify only the key and contents for the key
```

Comment:
Supposing I have `CREATE TABLE foo (foo_key VARCHAR) WITH (KEY='foo_key', ...);` and then I issue `INSERT INTO foo VALUES ("k");`. Are you saying this inserts a record with key `k` and value `NULL`? What if I wanted to insert a record with key `k` and value `k`? Is it possible for me to do that?

Reply:
Yeah... I definitely should flush this out a little more as well. My intention was that it is impossible to do what you want - and I think that's reasonable. If my key is a pointer to something in the value, then how can I have an empty value without an empty key?

Perhaps an option would be:
```sql
INSERT INTO foo (ROWKEY) VALUES ("k");
```
But I'm not sure that makes sense.

Reply:
Okay! after talking to @hjafarpour I understand how things work underneath the hood a little bit more and I'm inclined to say that a row schema should always include `ROWKEY`. This reflects how we handle CT statements today. For example:
```sql
CREATE TABLE bar (id VARCHAR, foo VARCHAR) WITH (key="id")
```
In this case, we expect kafka records to look like:

| key | value |
| -- | -- |
| id1 | id1, foo1|
| id2 | id2, foo2|

If they don't conform with that, then KSQL behavior is not well defined. So in this case, I think we should consider `ROWKEY` a valid column identifier. For example:
```sql
INSER INTO bar (ROWKEY, ID, FOO) VALUES ("k", null, null);
INSER INTO bar (ROWKEY, ID, FOO) VALUES ("k", "k", null);
INSER INTO bar (ROWKEY, ID, FOO) VALUES (null, "k", null);
INSER INTO bar (ROWKEY, ID, FOO) VALUES ("k", "x", null);
```
Would all be valid (though not necessarily sane)

Reply:
Thanks for kicking off this discussion! I updated the PR to also introduce `DELETE` semantics so that the two different workflows are distinct.